### PR TITLE
fix(dep_manifest, enrichment): 6 HIGH bugs from post-merge quorum review of #29

### PR DESCRIPTION
**Pull request by jsnyder**
## Summary

Six HIGH-severity bugs in the just-merged issue #29 work, fixed via TDD on this follow-up branch. Three were caught by the initial post-merge quorum review; three more were caught by a follow-up re-review pass on the changed files (two introduced by my own Bug-3 fix, one pre-existing in the same parser).

### Bugs fixed

| # | File | Bug | Commit |
|---|------|-----|--------|
| 1 | `context_enrichment.rs` | `parse_python_import("os, sys, json")` returned only `["os"]` (the `split_whitespace().next()` dropped the rest) | 4254843 |
| 2 | `dep_manifest.rs` | Empty/malformed pyproject.toml silently suppressed requirements.txt fallback | 0346872 |
| 3 | `dep_manifest.rs` | PEP 508 `name @ url` requirements were dropped entirely | 9389e83 |
| 4 | `dep_manifest.rs` | Bare URLs with `@` in authority (`https://user@host/...`) misparsed as named PEP 508 refs (regression from #3) | 624387a |
| 5 | `dep_manifest.rs` | Pip directives (`--find-links`, `--no-binary`, `-c`, `--pre`) leaked through as fake deps | 624387a |
| 6 | `dep_manifest.rs` | Local paths (`./dist/pkg.whl`, `../lib`) emitted as fake deps (pre-existing from #29, bundled because same parser) | b8c790f |

### Testing antipattern guard

Every test goes through `parse_dependencies(dir.path())` (production entry) or `normalize_import_to_dep_names("symbol: statement")` (production hydrated form), **not** the synthetic helper-input shape. This is the explicit lesson from the original-sin antipattern that shipped on #29: tests that used clean module paths like `"tokio::sync::Mutex"` passed while production hydration's `"Mutex: use tokio::sync::Mutex;"` silently no-op'd.

### Verification evidence

- 1396 unit tests pass (was 1393 pre-fix; 3 net-new test functions added covering 18 assertions). `cargo test --bin quorum`
- No clippy warnings on changed files. `cargo clippy --bin quorum --release`
- Final quorum re-review on the changed surface: 0 in-branch HIGH findings; 2 remaining HIGH are pre-existing and tracked in #75 and #82.
- Independent code review (subagent) verdict: ship after addressing one Important finding (PEP 508 environment markers + `@`-split). Pushback: the flagged input `foo @ ./local/path ; python_version < '3.10'` correctly extracts `"foo"` as the dep name — for Context7 enrichment we want every dep regardless of source URL or marker. Behavior is correct.

### Pre-existing HIGH findings filed as new issues

- #79 — `main.rs` `--source` + `--all` silently prefer `--all`
- #80 — `main.rs` doctor exit status from brittle stdout substring match
- #81 — `pipeline.rs` blocking semaphore `join()` can deadlock on current-thread Tokio
- #82 — `context_enrichment.rs` `query_docs` caches transient failures forever (older than #29; surfaced now)

### Original quorum review evidence

```
Pre-fix:  6 HIGH (3 in-branch + 3 pre-existing)
Post-fix: 2 HIGH (both pre-existing, both filed as issues)
```

## Test plan

- [x] `cargo test --bin quorum` (1396 pass, 0 fail)
- [x] `cargo clippy --bin quorum --release` (clean on touched files)
- [x] `cargo build --release --bin quorum` (success)
- [x] `quorum review src/{context_enrichment,dep_manifest}.rs` (in-branch HIGH cleared)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Context7 enrichment now analyzes project dependency manifests (Rust/Node/Python), selects up to 5 import-relevant deps, and applies curated or language-aware generic queries to surface targeted docs.
  * Shared enrichment fetcher used across multi-file reviews; fetch failures disable enrichment gracefully.

* **Improvements**
  * 24-hour LRU caching (positive and negative) with deterministic testing hook.
  * Per-file enrichment metrics collected and aggregated; new telemetry counters for resolved/resolve-failed/query-failed (backward-compatible).

* **Documentation**
  * Extensive design, rollout, and test plans covering parsing, precedence, edge cases, and telemetry.

### CODE REVIEW

File: docs/plans/2026-04-22-issue-29-context7-deps.md [hunk 1]
```diff
@@ -0,0 +1,1507 @@
...
+        "tokens_out": 0,
+        "duration_ms": 0,
+        "suppressed": 0
+    }"#;
+    let entry: TelemetryEntry = serde_json::from_str(old)
+        .expect("old JSONL rows must deserialize after schema bump");
+    assert_eq!(entry.context7_resolved, 0);
+    assert_eq!(entry.context7_resolve_failed, 0);
+    assert_eq!(entry.context7_query_failed, 0);
+}
```
[the remark this answers is not in the document]
Comment (jsnyder) [reply docs/plans/2026-04-22-issue-29-context7-deps.md:1351]: Pushing back as FP — this fixture is for `TelemetryEntry` (`~/.quorum/telemetry.jsonl`), not `ReviewRecord` (`~/.quorum/reviews.jsonl`). Two different files, two different schemas.

Real `TelemetryEntry` shape (from `src/telemetry.rs` head-of-branch):
```rust
pub struct TelemetryEntry {
    pub ts: DateTime<Utc>,
    pub files: Vec<String>,
    pub findings: HashMap<String, usize>,
    pub model: String,
    pub tokens_in: u64,
    pub tokens_out: u64,
    pub duration_ms: u64,
    pub suppressed: usize,
    #[serde(default)] pub context7_resolved: u32,
    #[serde(default)] pub context7_resolve_failed: u32,
    #[serde(default)] pub context7_query_failed: u32,
}
```

The fixture's `ts/files/findings/suppressed` keys match this exactly. The `run_id/timestamp/repo/invoked_from/files_reviewed/findings_by_severity/tokens_cache_*` schema you cited belongs to `ReviewRecord` in `src/review_log.rs:233`, which lives in `reviews.jsonl` — a separate telemetry file with no Context7 counters.

CR1 (the CLAUDE.md fix) was real and is now committed in 95b320e — it clarifies that the context7_* counters live on `TelemetryEntry`, not `ReviewRecord`. CR2 is the same conflation in reverse.

No change needed.

**Review by jsnyder — COMMENTED** [review]
# Independent Code Review (Phase 8 — superpowers:code-reviewer)

## Overall Assessment

This is a **strong, mature PR** that demonstrates the value of multiple review passes layered on disciplined TDD. The code shows obvious care: every fix has a regression test with a comment naming the bug class (e.g. "Bug 3", "antipatterns reviewer MUST-FIX 3.1", "Quorum re-review HIGH"). The orchestration boundary between `dep_manifest` (pure parsers) and `context_enrichment` (network + cache + policy) is clean. Telemetry counters are correctly summed across files in `main.rs:1013-1015` and use `serde(default)` for backward compat.

### Strengths
- **Test density is exceptional.** `dep_manifest.rs` has ~30 cases pinning every parser branch, including subtle precedence rules (PEP 621 empty array vs Poetry, wrong-type deps, named-direct-URL refs vs `@`-in-authority URLs).
- **LRU eviction tests** (`src/context_enrichment.rs:1323` and `:1371`) directly assert the regression — "old `clear()` would push this to 6" — which is exactly how a regression test should read.
- **Negative-result TTL caching** with injectable clock is a clean design and the test at `:1130` exercises the TTL boundary deterministically.
- **Cache-key correctness** for `query_docs` uses `(library_id, query, max_tokens)`.
- `try_fetch_one` correctly inserts into `seen` *unconditionally* (`src/context_enrichment.rs:483-487`) so resolve-fail or query-fail can't double-count when the same name appears in both the import-matched loop and the curated-frameworks loop.

---

## Findings

### Critical
None.

### Important

**I1 — `Box::leak` is bounded but the leak repeats per `quorum daemon` review.**
`src/main.rs:629` leaks one `Context7HttpFetcher` per `run_review` invocation. For one-shot CLI use this is fine (process exits). For `quorum daemon --watch-dir` and the MCP `serve` path, every review through `main::run_review` would add another leaked fetcher (~hundreds of bytes + reqwest client internals + a TLS connection pool). Over weeks of daemon uptime this is a small but unbounded growth. Two fixes either work:
1. Cache the leaked fetcher in a `OnceLock<&'static Context7HttpFetcher>` so the leak happens at most once per process.
2. Convert `ContextFetcher`'s required bound from `&'a` to `Arc<dyn ContextFetcher + Send + Sync>` end-to-end and drop `Box::leak` entirely. `CachedContextFetcher<'a>` is the only thing forcing the borrow.

The comment at `src/main.rs:619-621` already identifies the per-invocation leak; I'd just guard it for the daemon/MCP cases that route through this same function.

**I2 — Per-file fallback at `src/pipeline.rs:436` and `:781` builds a fresh `Context7HttpFetcher` *and* fresh cache per file when `pipeline_config.context7_fetcher` is `None`.**
The shared-cache path in `main.rs` is wired, but tests, the MCP `serve` handler, and any direct `pipeline_config` consumer will silently take the fallback and re-pay TLS handshake + cold cache per file. Consider logging `tracing::warn!` once when the fallback path is taken with `>1` file in the run, so silent re-hammering is visible. Or fail-fast in the >1-file case.

### Minor

**M1 — `parse_python_import` strips `#` comments via `split('#').next()`.**
Python identifiers can't contain `#`, so this is fine in practice — but the more correct form `split_once(" #")` (require whitespace before the comment) is identical for valid Python and clearer in intent. Optional.

**M2 — `parse_rust_use` only returns the first non-empty `::` segment, which is correct for top-level crates but drops `crate::`, `self::`, `super::` qualifiers as if they were external crate names.**
A `use crate::foo::Bar` will yield a dep-name lookup for `"crate"`, which never matches a real Cargo dep. That's harmless (the import-filter just won't find a dep), but emitting `"crate"`/`"self"`/`"super"` into `import_targets` does pollute `build_code_aware_query`'s keyword list (it filters by `len() > 2` so `"crate"` slips through and would appear as a query keyword for some other matched dep). Recommend adding `if matches!(head, "crate" | "self" | "super") { return Vec::new(); }` to `parse_rust_use`.

**M3 — `parse_pyproject` "wrong-type dependencies" branch returns `Some(Vec::new())` and emits a warning, which is a defensible call but undocumented in the Context7 enrichment section of `CLAUDE.md`.**
Worth a one-line note so future maintainers don't re-add the fall-through.

**M4 — `parse_package_json` does no dedupe across `dependencies` / `devDependencies` / `peerDependencies` / `optionalDependencies`.**
Cargo and pyproject also don't dedupe (relying on the orchestrator). Consistent, but means the same name can appear up to 4× in the returned `Vec<Dependency>`. Not a correctness bug — `enrich_for_review`'s `import_matched` filter dedupes by name — but slightly wasteful in the per-file loops. Optional.

### Nit

**N1 — `cargo_dep_in_both_dependencies_and_dev_dependencies_appears` test asserts `count >= 1` but the underlying logic uses a `seen` `HashSet` so the count is exactly 1.** Tighten the assertion or update the comment ("currently appends both") which now contradicts the implementation.

**N2 — `format_context_section` (`src/context_enrichment.rs:344`) sanitizes triple backticks to `'''`, which is consistent with prior trust-boundary work** (#67), but the language tag on the fence is empty. Setting it to `text` or the doc's known language helps downstream Markdown renderers.

**N3 — `CachedContextFetcher::new_with_clock` `.expect("max_entries.max(1) is non-zero")` is correct but the `expect` message is unnecessary** — the `unwrap()` would never panic. A short `// SAFETY:` comment or `unsafe { NonZeroUsize::new_unchecked(...) }` would make intent clearer.

---

## Merge Recommendation

**Approve with minor follow-ups.** The critical correctness work has been done thoroughly. Suggest merging as-is and filing follow-up issues for:
- I1 (bounded `Box::leak` for daemon/MCP) — small risk, but worth tracking.
- I2 (visibility when the per-file fallback path runs in multi-file reviews).
- M2 (`parse_rust_use` skipping `crate`/`self`/`super`) — quick fix, prevents query pollution.

The architecture, test coverage, and documentation discipline on this PR are above the bar for the codebase. The acknowledged out-of-scope list (#75/#76/#82/#84) is reasonable scoping.

File: src/dep_manifest.rs [hunk 2]
```diff
@@ -0,0 +1,806 @@
...
+    }
+    let poetry_table = parsed
+        .get("tool")
+        .and_then(|t| t.get("poetry"))
+        .and_then(|p| p.get("dependencies"))
+        .and_then(|d| d.as_table());
+    let Some(table) = poetry_table else {
+        // Neither PEP 621 nor Poetry section recognized — let requirements.txt
+        // handle this project.
+        return None;
```
[the remark this answers is not in the document]
Comment (jsnyder) [reply src/dep_manifest.rs]: Fixed in 69a9e78 — applied your suggested diff almost verbatim. Treats wrong-type `[tool.poetry.dependencies]` as explicitly empty with a `tracing::warn` naming the offending TOML kind, mirroring the existing `[project].dependencies` guard. Two new tests pin the wrong-type and the no-deps-key behaviors.

File: src/main.rs [hunk 3]
```diff
@@ -611,6 +612,32 @@ fn run_review(opts: cli::ReviewOpts) -> i32 {
...
+                    let cached = crate::context_enrichment::CachedContextFetcher::new(leaked, 32);
+                    Some(std::sync::Arc::new(cached)
+                        as std::sync::Arc<dyn crate::context_enrichment::ContextFetcher>)
+                }
+                Err(e) => {
+                    tracing::warn!(error = %e, "Context7 HTTP fetcher init failed; per-file ad-hoc fetcher will be used");
+                    None
+                }
+            }
+        };
```
[the remark this answers is not in the document]
Comment (jsnyder) [reply src/main.rs]: Fixed in 9424ba8 with a slight design tweak: instead of carrying `context7_init_failed` only in a tuple, I added a sticky `context7_disabled: bool` field to `PipelineConfig` and a `context7_skip_reason()` predicate that consolidates both `--skip-context7` and bootstrap-failure into a single guard at the top of each pipeline path. Same observable behavior as your suggestion; one extra unit test per branch (3 total). The per-file ad-hoc fallback path is preserved for tests / MCP / http_server consumers that don't wire a shared fetcher.

File: src/context_enrichment.rs [hunk 4]
```diff
@@ -15,28 +15,277 @@ pub trait ContextFetcher: Send + Sync {
...
+    let stmt = stmt.trim().trim_end_matches(';');
+    // Skip empty leading segments so absolute paths (`use ::tokio::...`) work
+    // the same as relative ones.
+    let head = stmt
+        .split("::")
+        .find(|s| !s.is_empty())
+        .unwrap_or("")
+        .trim();
+    if head.is_empty() { return Vec::new(); }
+    vec![head.to_string()]
```
[the remark this answers is not in the document]
Comment (jsnyder) [reply src/context_enrichment.rs]: Fixed in 2439e79. Added an outer-brace check to `parse_rust_use`: when the stmt starts with `{` and ends with `}`, strip braces, split on `,`, and run the existing first-non-empty-`::`-segment logic per group member. Inner-grouped form (`use serde::{A, B}`) stays on the cheap single-pass path. Three regression tests pin the outer-grouped, per-segment-absolute-path, and degenerate single-member cases.

**Comment by jsnyder** [comment]
## Update — review feedback addressed

Pushed 4 commits addressing both CodeRabbit's new round and the independent code-reviewer's findings.

**CodeRabbit (new round):**
- **CR6 — grouped Rust `use { ... }` imports** → fixed in 2439e79. `parse_rust_use` now detects outer braces, splits on commas, and applies first-non-empty-`::` per group member. Inner-grouped form (`use serde::{A, B}`) stays on the cheap single-pass path.
- **CR7 — malformed `[tool.poetry.dependencies]`** → fixed in 69a9e78. Mirrors the existing `[project].dependencies` wrong-type guard: warn + return `Some(empty)` instead of falling through to `requirements.txt`.
- **CR8 — disable Context7 on bootstrap failure** → fixed in 9424ba8. Added sticky `context7_disabled: bool` to `PipelineConfig` and a `context7_skip_reason()` predicate that consolidates `--skip-context7` and bootstrap-failure into a single guard.

**Independent reviewer (Phase 8):**
- **N1 — cargo dedupe assertion** → tightened to `==1` (matches actual behavior since the target.* deps fix in 0aca79c). Commit a257810.
- **N2 — empty fence language** → switched to ```text``` for explicit "opaque blob" semantics. Commit a257810.
- **M3 — pyproject wrong-type doc** → CLAUDE.md updated with both the precedence rule and the CR8 bootstrap-failure flow. Commit a257810.
- **I1 — `Box::leak` per-call** → triaged as moot. `run_review` is the one-shot CLI dispatcher (`main()` calls `std::process::exit` immediately after); the long-lived `daemon`/`serve` paths use their own pipelines. Comment updated to document the once-per-process guarantee. Commit a257810.
- **M2 — `parse_rust_use` `crate`/`self`/`super` pollution** → triaged as FP. The reviewer's pollution path was wrong: `extract_query_keyword` uses the symbol prefix (`Mutex` from `"Mutex: use crate::foo;"`), not parse_rust_use's output. Existing test `normalize_local_paths_yield_keyword_that_wont_match_real_deps` already pins the harmless behavior.
- **I2 — per-file fallback observability** → not applied. Per-file fallback is intentional for tests / MCP / http_server consumers that don't wire a shared fetcher; the CR8 fix removes the broken main.rs path.
- **M1, M4, N3** — cosmetic preferences, skipped.

**Test status:** 1414/1414 passing.
**Quorum feedback recorded:** 7 verdicts (5 TP + 2 FP) for calibrator training.
